### PR TITLE
[TRA-64] Use market specific insurance fund for cross or isolated markets

### PR DESCRIPTION
**Pull request by shrenujb**
### Changelist
Add APIs to find which insurance fund to use based on the perpetual traded
Modify usages of Insurance Fund to go through this API

### Test Plan
[Describe how this PR was tested (if applicable)]

### Author/Reviewer Checklist
- [ ] If this PR has changes that result in a different app state given the same prior state and transaction list, manually add the `state-breaking` label.
- [ ] If the PR has breaking postgres changes to the indexer add the `indexer-postgres-breaking` label.
- [ ] If this PR isn't state-breaking but has changes that modify behavior in `PrepareProposal` or `ProcessProposal`, manually add the label `proposal-breaking`.
- [ ] If this PR is one of many that implement a specific feature, manually label them all `feature:[feature-name]`.
- [ ] If you wish to for mergify-bot to automatically create a PR to backport your change to a release branch, manually add the label `backport/[branch-name]`.
- [ ] Manually add any of the following labels: `refactor`, `chore`, `bug`.

### CODE REVIEW

File: protocol/x/perpetuals/keeper/perpetual.go [hunk 1]
```diff
@@ -29,6 +31,29 @@ import (
 	gometrics "github.com/hashicorp/go-metrics"
 )
 
+func (k Keeper) GetInsuranceFundName(ctx sdk.Context, perpetualId uint32) (string, error) {
```
Comment (shrenujb) [inline protocol/x/perpetuals/keeper/perpetual.go:34]: Add function comments for both

File: protocol/x/clob/keeper/deleveraging.go [hunk 2]
```diff
@@ -132,18 +132,33 @@ func (k Keeper) MaybeDeleverageSubaccount(
 
 // GetInsuranceFundBalance returns the current balance of the insurance fund (in quote quantums).
 // This calls the Bank Keeper’s GetBalance() function for the Module Address of the insurance fund.
-func (k Keeper) GetInsuranceFundBalance(
-	ctx sdk.Context,
-) (
-	balance *big.Int,
-) {
+func (k Keeper) GetInsuranceFundBalance(ctx sdk.Context, perpetualId uint32) (balance *big.Int) {
```
Comment (vincentwschau) [inline protocol/x/clob/keeper/deleveraging.go:135]: nit: Update comment to note that this gets the balance for the insurance fund for a specific perpetual.

File: protocol/x/subaccounts/keeper/transfer_test.go [hunk 3]
```diff
@@ -673,27 +674,47 @@ func TestTransferInsuranceFundPayments(t *testing.T) {
...
 		},
 		"success - can send zero payment": {
+			perpetual:                           constants.BtcUsd_SmallMarginRequirement,
 			insuranceFundBalance:                2500,
 			subaccountModuleAccBalance:          600,
 			quantums:                            big.NewInt(0),
 			expectedSubaccountsModuleAccBalance: 600,
 			expectedInsuranceFundBalance:        2500,
 		},
+		"success - send to isolated insurance fund account": {
```
Comment (vincentwschau) [inline protocol/x/subaccounts/keeper/transfer_test.go:700]: nit: Can you also add a test-case for failure to send from the insurance fund for an isolated market due to lacking funds?

File: protocol/x/perpetuals/keeper/perpetual.go [hunk 4]
```diff
@@ -29,6 +31,33 @@ import (
...
+	perpetual, err := k.GetPerpetual(ctx, perpetualId)
+	if err != nil {
+		return "", err
+	}
+
+	if perpetual.Params.MarketType == types.PerpetualMarketType_PERPETUAL_MARKET_TYPE_ISOLATED {
+		return types.InsuranceFundName + ":" + lib.UintToString(perpetualId), nil
+	} else if perpetual.Params.MarketType == types.PerpetualMarketType_PERPETUAL_MARKET_TYPE_CROSS {
+		return types.InsuranceFundName, nil
+	}
```
Comment (vincentwschau) [inline protocol/x/perpetuals/keeper/perpetual.go:47]: Question, is our cosmos-sdk version up to date enough for this [code snippet](https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-028-public-key-addresses.md#module-account-addresses) to work?

File: protocol/testutil/constants/perpetuals.go [hunk 5]
```diff
@@ -299,8 +299,71 @@ var (
...
+			MarketId:          uint32(2),
+			AtomicResolution:  int32(-9),
+			DefaultFundingPpm: int32(0),
+			LiquidityTier:     uint32(3),
+			MarketType:        perptypes.PerpetualMarketType_PERPETUAL_MARKET_TYPE_CROSS,
+		},
+		FundingIndex: dtypes.ZeroInt(),
+	},
+	{
+		Params: perptypes.PerpetualParams{
```
[the remark this answers is not in the document]
Comment (shrenujb) [reply protocol/testutil/constants/perpetuals.go:354]: The constant above is for SOL which is a cross market vs this one which is meant to represent an isolated market. See the difference in market types

File: protocol/x/perpetuals/keeper/perpetual.go (hunk 4 above)
Comment (shrenujb) [reply protocol/x/perpetuals/keeper/perpetual.go:47]: Are you referring to the address.Module API?
It should be considering we already use authtypes.NewModuleAddress which uses that underneath

File: protocol/testutil/constants/perpetuals.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vincentwschau) [reply protocol/testutil/constants/perpetuals.go:354]: I meant this [constant](https://github.com/dydxprotocol/v4-chain/pull/1132/files#diff-a5b97f4ea9ad77e38d65c3831bf9f94aaf97352e6e825d73f9f7ad14fb737209R302).

File: protocol/x/clob/keeper/deleveraging.go [hunk 6]
```diff
@@ -130,20 +130,36 @@ func (k Keeper) MaybeDeleverageSubaccount(
...
+
+func (k Keeper) GetCrossInsuranceFundBalance(ctx sdk.Context) (balance *big.Int) {
 	usdcAsset, exists := k.assetsKeeper.GetAsset(ctx, assettypes.AssetUsdc.Id)
 	if !exists {
 		panic("GetInsuranceFundBalance: Usdc asset not found in state")
 	}
 	insuranceFundBalance := k.bankKeeper.GetBalance(
 		ctx,
-		types2.InsuranceFundModuleAddress,
+		perptypes.InsuranceFundModuleAddress,
```
[the remark this answers is not in the document]
Comment (vincentwschau) [reply protocol/x/clob/keeper/deleveraging.go:162]: nit: The suggestion to update the error message for the panic is good.

File: protocol/x/perpetuals/keeper/perpetual.go (hunk 4 above)
Comment (shrenujb) [reply protocol/x/perpetuals/keeper/perpetual.go:47]: I don't think the 2 solutions would be very different in terms of impact (having separate addresses for these accounts) and the latter would require a bigger refactor

**Review by vincentwschau — APPROVED** [review]

**Pull request merged**